### PR TITLE
Assignment3

### PR DESCRIPTION
**Pull request by MikePHaynes**

### CODE REVIEW

**Comment by sammyboy2000** [comment]
I will review this.

**Comment by sammyboy2000** [comment]
## Assignment

For this assignment do the following:

- Create *two* interfaces. One to represent the `JokeService` and one interface to represent outputting the joke to the screen (`Console.WriteLine`). ✔
- The `JokeService` will need to have the interface applied to it. ✔
- Create an implementation of the output interface that writes out the joke to the console. ✔
- Implement the `Jester` class. It should take in both interfaces as dependencies. These dependencies should be null checked. ❌
- The `Jester` class `TellJoke()` method should retrieve a joke from the `JokeService`. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
- Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ❌
- Enable nullability in the .csproj files and ensure no errors or warnings: ✔

## Fundamentals
- Ensure you enable:
  - nullable reference types is enabled  ✔
  - net6 targetted  ✔
  - C# 10.0  ✔
  - and enabled .NET analyzers for both projects ✔
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ✔
- All of the above should be unit tested ❌
- Choose simplicity over complexity ✔
  
## Extra Credit
- Unit test your implementation that writes the joke out to the screen. How hard could it be to unit test a single line method ;)? ❌
- The Geek jokes API that is being used can also return jokes in a JSON format. Update the `JokeService` to retrieve jokes using JSON. `GetJoke` should still return a string. ❌

Unit tests still need to be written.
 Jester's interfaces need to be null checked.

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
+
+namespace CanHazFunny
+{
+    public class Jester
+    {
+        private readonly IJokeOutput _jokeOutput;
+        private readonly IJokeService _jokeService;
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:8]: Consider making these auto properties

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        public Jester(IJokeOutput? jokeOutput, IJokeService? jokeService)
+        {
+            _jokeOutput = jokeOutput ?? throw new ArgumentNullException(nameof(jokeOutput));
+            _jokeService = jokeService ?? throw new ArgumentNullException(nameof(jokeService));
+        }
+
+        public void TellJoke()
+        {
+            string joke = _jokeService.GetJoke();
+            while (joke.Contains("Chuck") || joke.Contains("Norris")) joke = _jokeService.GetJoke();
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:19]: This is a personal preference, but I like to wrap the body in brackets just for readability
```suggestion
            while (joke.Contains("Chuck") || joke.Contains("Norris")) { joke = _jokeService.GetJoke(); }
```

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 3]
```diff
@@ -1,10 +1,45 @@
...
+        {
+            new Jester(new JokeOutput(), null);
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentNullException))]
+        public void Constructor_ServiceDependencyNull_ThrowsException()
+        {
+            new Jester(null, new JokeService());
+        }
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:22]: Do you have these test names backwards?

**Pull request closed without merging**